packages/next/src/server/app-render/app-render-prerender-utils.ts

```diff
@@ -0,0 +1,226 @@
+// in one that doesn't close even when the underlying is complete.
+export class ReactServerResult {
+  private _stream: null | ReadableStream<Uint8Array>
+
+  constructor(stream: ReadableStream<Uint8Array>) {
+    this._stream = stream
+  }
+
+  tee() {
+    if (this._stream === null) {
+      throw new Error(
+        'Cannot tee a ReactServerResult that has already been consumed'
+      )
+    }
+    const tee = this._stream.tee()
+    this._stream = tee[0]
+    return tee[1]
+  }
+
+  consume() {
+    if (this._stream === null) {
+      throw new Error(
+        'Cannot consume a ReactServerResult that has already been consumed'
+      )
+    }
+    const stream = this._stream
+    this._stream = null
+    return stream
+  }
+}
+
+type ReactPromise<T> = Promise<T> & {
+  status?: string
+  value?: T
+  reason?: unknown
+}
+export type ReactServerPrerenderResolveToType = {
+  prelude: ReadableStream<Uint8Array>
+}
+
+export async function createReactServerPrerenderResult(
+  underlying: Promise<ReactServerPrerenderResolveToType>
+): Promise<ReactServerPrerenderResult>
+export async function createReactServerPrerenderResult(
+  underlying: ReactPromise<ReactServerPrerenderResolveToType>
+): Promise<ReactServerPrerenderResult> {
+  underlying.catch(() => {})
+
+  const chunks: Array<Uint8Array> = []
+  const { prelude } = await underlying
+  await new Promise((resolve, reject) => {
+    const reader = prelude.getReader()
+    function progress(result: ReadableStreamReadResult<Uint8Array>) {
+      if (result.done) {
+        resolve(chunks)
+        return
+      } else {
+        chunks.push(result.value)
+        reader.read().then(progress, error)
+      }
+    }
+    function error(e: unknown) {
+      reader.cancel(e)
+      reject(e)
+    }
+    reader.read().then(progress, error)
+  })
+  return new ReactServerPrerenderResult(chunks)
+}
+
+export async function createReactServerPrerenderResultFromRender(
+  underlying: ReadableStream<Uint8Array>
+): Promise<ReactServerPrerenderResult> {
+  const chunks: Array<Uint8Array> = []
+  await new Promise((resolve, reject) => {
+    const reader = underlying.getReader()
+    function progress(result: ReadableStreamReadResult<Uint8Array>) {
+      if (result.done) {
+        resolve(chunks)
+        return
+      } else {
+        chunks.push(result.value)
+        reader.read().then(progress, error)
+      }
+    }
+    function error(e: unknown) {
+      reader.cancel(e)
+      reject(e)
+    }
+    reader.read().then(progress, error)
+  })
+  return new ReactServerPrerenderResult(chunks)
+}
+export class ReactServerPrerenderResult {
+  private _chunks: null | Array<Uint8Array>
+
+  private assertChunks(expression: string): Array<Uint8Array> {
+    if (this._chunks === null) {
+      throw new Error(
+        `Cannot \`${expression}\` on a ReactServerPrerenderResult that has already been consumed`
```

Comment:
maybe a good usecase for [`InvariantError`](https://github.com/vercel/next.js/blob/canary/packages/next/src/shared/lib/invariant-error.ts)? bunch of other places here as well